[ty] support PEP 613 type aliases

carljm opened the pull request:
Refs https://github.com/astral-sh/ty/issues/544

## Summary

Takes a more incremental approach to PEP 613 type alias support (vs https://github.com/astral-sh/ruff/pull/20107). Instead of eagerly inferring the RHS of a PEP 613 type alias as a type expression, infer it as a value expression, just like we do for implicit type aliases, taking advantage of the same support for e.g. unions and other type special forms.

The main reason I'm following this path instead of the one in https://github.com/astral-sh/ruff/pull/20107 is that we've realized that people do sometimes use PEP 613 type aliases as values, not just as types (because they are just a normal runtime assignment, unlike PEP 695 type aliases which create an opaque `TypeAliasType`).

This PR doesn't yet provide full support for recursive type aliases (they don't panic, but they just fall back to `Unknown` at the recursion point). This is probably post-beta work.

## Test Plan

Added mdtests.

Many new ecosystem diagnostics, probably mostly just because we understand new types in lots of places, but I'll spot-check for potential problems.

Conformance suite changes are correct.

Performance regression is also just due to understanding lots of new types; nothing we do in this PR is inherently expensive.

sharkdp commented:
> But there are quite a few `invalid-type-form` resulting from failing to understand stringified annotations in these aliases. It looks like it may not be a good idea to land this without adding some kind of support or workaround for those.

Given that this PR implements PEP 613 aliases as if they were implicit aliases, a simple first-iteration version might be to add a new `Type::LiteralString(_) => todo_type!("Stringified annotation in type alias")` branch in `Type::in_type_expression`?

MichaReiser commented:
> Performance regression is also just due to understanding lots of new types; nothing we do in this PR is inherently expensive.

While true, I think it's still worth looking at the performance regression to see if anything stands out. 50% is way more than even the type-of-self PR introduced and I suspect what we're seeing is some bad run-away, similar to what we've seen with pydantic the other day. I'm not suggesting that we necessearily need to fix the performance regression, but we should at least try to understand where it's coming from and if there's something that clearly stands out.

carljm commented:
Adding support for stringified RHS eliminated the false positives from that -- but just added a bunch more diagnostics from understanding even more type aliases!

Some major categories of new diagnostics that I'm seeing:
1. `invalid-type-form` due to seeing a type of `object` or `UnionType` in a type expression. It looks like these are mostly due to specialized generic aliases or callable types or other things we don't understand yet. Probably need to either delay this PR until after support for those lands, or find a way to silence the false positives.
2. A ton of new `invalid-key` diagnostics from pydantic, because we now understand more type aliases for unions of typed dicts, and we don't support tagged-union discrimination. (We will also need to support aliased conditional expressions to get rid of some of these, because in some places pydantic does `schema_type = schema['type']` and then performs checks on the `schema_type` variable.) These we just have to live with until we implement those features.

Trying to debug the first category is painful, because you just get this opaque error at the usage site of the type alias and have to dig back to its definition and then figure out what we didn't understand, sometimes several layers of type aliases deep. I'm a bit worried that even once we support all the "correct" forms of type aliases, users will still get confusing and unhelpful error messages if they write bad type aliases. Maybe one way we could mitigate this, at least for `TypeAlias` if not for implicit type aliases, would be to always create a known-instance union type that is valid in type expressions for any vertical-pipe operators, even if some of the components weren't valid/known.

AlexWaygood commented:
The pydantic benchmark is panicking FYI, so the codspeed results aren't up to date:

```
  Expected between 1 and 5000 diagnostics on project 'pydantic' but got 6334
```

carljm commented:
Added a Todo type for subscripting `UnionType` known instance, and added specialized handling for PEP 604 unions in a `TypeAlias`, and that looks a lot better. Still a few added `invalid-type-form` diagnostics to debug.

AlexWaygood commented on crates/ty_python_semantic/resources/mdtest/pep613_type_aliases.md:
```diff
@@ -1,8 +1,124 @@
+MyAlias: TypeAlias = int | Callable[[str], int]
+
+def _(x: MyAlias):
+    # TODO: int | (str) -> int
```
This todo seems done?

```suggestion
```

MichaReiser commented:
@carljm would you mind updating https://github.com/astral-sh/ty/issues/445, now that type aliases are supported?
